### PR TITLE
drop outstanding_requests lock before sending repair requests

### PR DESCRIPTION
**Pull request by jbiseda**
- misc turbine metrics
- drop outstanding_requests lock before sending repair request packets

### CODE REVIEW

File: core/src/repair_service.rs [hunk 1]
```diff
@@ -262,29 +268,51 @@ impl RepairService {
...
+                        &repair_info.repair_validators,
+                        &mut outstanding_requests,
+                    ) {
+                        Some((req, to))
+                    } else {
+                        None
+                    }
+                })
+                .collect();
+            drop(outstanding_requests);
```
Comment (behzadnouri) [inline core/src/repair_service.rs:290]: I think you can utilize `?` operator here. Also giving `outstanding_requests` a scope will avoid manual `drop`. So something like:
```rust
let batch: Vec<_> = {
  let mut outstanding_requests = outstanding_requests.write().unwrap();
  repairs
    .into_iter()
    .filter_map(|repair_request| {
        let (to, req) = serve_repair.repair_request(...).ok()?;
        Some((req, to))
    })
    .collect()
};
```

File: core/src/window_service.rs [hunk 2]
```diff
@@ -39,6 +40,34 @@ use std::{
 pub type DuplicateSlotSender = CrossbeamSender<Slot>;
 pub type DuplicateSlotReceiver = CrossbeamReceiver<Slot>;
 
+#[derive(Default)]
+pub struct WindowServiceMetrics {
+    pub run_insert_count: u64,
+    pub num_shreds_received: u64,
+    pub shred_receiver_elapsed_us: u64,
+    pub prune_shreds_elapsed_us: u64,
+}
```
Comment (behzadnouri) [inline core/src/window_service.rs:49]: none seem to need `pub`

File: ledger/src/blockstore.rs [hunk 3]
```diff
@@ -156,7 +156,7 @@ pub struct Blockstore {
     slots_stats: Arc<Mutex<SlotsStats>>,
 }
 
-struct SlotsStats {
+pub struct SlotsStats {
```
Comment (behzadnouri) [inline ledger/src/blockstore.rs:159]: why `pub`?

**Review by behzadnouri — APPROVED** [review]
LGTM

**Review by sakridge — APPROVED** [review]
lgtm

**Review by carllin — DISMISSED** [review]

File: core/src/window_service.rs (hunk 2 above)
Comment (jbiseda) [reply core/src/window_service.rs:49]: removed pub

File: ledger/src/blockstore.rs (hunk 3 above)
Comment (jbiseda) [reply ledger/src/blockstore.rs:159]: removed

File: core/src/repair_service.rs (hunk 1 above)
Comment (jbiseda) [reply core/src/repair_service.rs]: That's more succinct, changed.

**Pull request merged**